programs/self-extracting/CMakeLists.txt

```diff
@@ -0,0 +1,5 @@
+add_custom_target (self-extracting ALL
+    ${CMAKE_BINARY_DIR}/utils/self-extracting-executable/compressor clickhouse ../clickhouse
+    DEPENDS clickhouse compressor
```

Comment:
This will not replace clickhouse with compressed file (it will simply create new binary under `self-extracting` binary folder), so it does not resolve the initial issue, or am I missing something?

Plus it does not looks correct to provide such binaries in packages (deb and others), packages are compressed anyway, and decompressing them on first run does not worth it.

Oh, and by the way, it will report an error if file already exists (i.e. running build multiple times):

```
FAILED: programs/self-extracting/CMakeFiles/self-extracting /bld/programs/self-extracting/CMakeFiles/self-extracting 
cd /bld/programs/self-extracting && /bld/utils/self-extracting-executable/compressor clickhouse ../clickhouse
Error: output file [clickhouse] already exists
```

Reply:
I think it's better to play safe and preserve original file, so decided to create compressed in separate directory.

Reply:
@azat where do we build packages? so I could exclude compressed...

Reply:
>@azat where do we build packages? so I could exclude compressed...

packages uses binaries that was built.
Right now nothing changed, this binary is simply generated in a separate folder and ignored, so packages has not been changed.

Reply:
@azat deb, rpm, and tarballs also should use self-extracting binary. 
And we can set the compression level of the packages to zero.

The motivation is the following:
- deb packages cannot use a modern compression library (other than gzip), because we must support very old Linux distributions;
- we need to use exactly the same binary in single-binary package and in other packages to avoid potential inconsistencies;
- we want to package clickhouse-odbc-bridge and clickhouse-library-bridge into the single self-extracting executable to make single-binary setups full-featured (and, consequently, there will be fewer reasons to use .deb or .rpm and absolutely zero reasons to use .tar.gz).
- ideally, we want to ditch the "-dbg" package and always include the debug info.

Reply:
I have another viewpoint on the topic.

> deb packages cannot use a modern compression library (other than gzip), because we must support very old Linux distributions;

That's true, for sure

> we need to use exactly the same binary in single-binary package and in other packages to avoid potential inconsistencies;

And it's 100% correct as well

> we want to package clickhouse-odbc-bridge and clickhouse-library-bridge into the single self-extracting executable to make single-binary setups full-featured (and, consequently, there will be fewer reasons to use .deb or .rpm and absolutely zero reasons to use .tar.gz);

Would it mean they will be extracted on the fly after/at the install step? It's an antipattern for packages. Nothing should change the data on the disk after unpacking.

> parallel compression and modern compression library will make packages build faster (right now the packages wrapping can take longer than ClickHouse build itself - several minutes).

It's a past year statement. Now all packages are ready in 30 seconds, from [here](https://github.com/ClickHouse/ClickHouse/runs/7179577626?check_suite_focus=true#step:6:9744) to [here](https://github.com/ClickHouse/ClickHouse/runs/7179577626?check_suite_focus=true#step:6:10059)

And the main question is, what about docker? What is the plan to put it there?

Reply:
> And the main question is, what about docker? What is the plan to put it there?

We can keep extracted binaries inside Docker - the state after package install (to speed up cold start).

> Would it mean they will be extracted on the fly after/at the install step? It's an antipattern for packages. Nothing should change the data on the disk after unpacking.

On the install step.

> It's an antipattern for packages.

Maybe break the rule? Or we can have the main binary compressed (self-extracting) and bridges as is.

Reply:
we had some discussion today on sync - we can't install compressed binary into /usr/bin - it will try to self-extract here under user's credentials on first run
and I'm not sure it's just an antipattern - I guess package manager can rely on files' signatures to manage packages

Reply:
But we run `clickhouse install` on package install, so it is run as root.

Reply:
> On the install step.
> Maybe break the rule? Or we can have the main binary compressed (self-extracting) and bridges as is.

It's the basic packages' rule. If we are going to violent it, it's better to just drop packages completely, unfortunately.

Files that are unpacked from packages must remain the same because package managers have integrity checks for installed files:

```
> rpm -ql --dump clickhouse-client-22.6.2.12.noarch.rpm
error: cannot open Packages database in /var/lib/rpm
error: cannot open Packages database in /var/lib/rpm
/etc/clickhouse-client/config.xml 2273 1656420967 b3bccf6987b8100985387391bdb803fa56a8fc632855286ea6d4292379f11dad 0100644 root root 1 0 1 X
/usr/bin/clickhouse-benchmark 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/bin/clickhouse-client 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/bin/clickhouse-compressor 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/bin/clickhouse-format 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/bin/clickhouse-local 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/bin/clickhouse-obfuscator 10 1656422127 0000000000000000000000000000000000000000000000000000000000000000 0120000 root root 0 0 1 clickhouse
/usr/share/doc/clickhouse-client/AUTHORS 203 1656420961 a958439a16a73c4fb22abec7fcd6efb8fc04267a1ab3af866e9f31a91555e1e2 0100664 root root 0 0 1 X
/usr/share/doc/clickhouse-client/CHANGELOG.md 221346 1656420961 4435f21c28615ef291f006ffbeae71764307d872ca96d486d682099787dc5337 0100664 root root 0 0 1 X
/usr/share/doc/clickhouse-client/LICENSE 11389 1656420961 efbf644e6c8898b2b0320e27c1ddfa1a556efcd64646da82e0db8c01ba0604c2 0100664 root root 0 0 1 X
/usr/share/doc/clickhouse-client/README.md 1458 1656420961 de14ffea4595ac64676c9e143c07819bb4297b2641d6f88e4c8668f35805c3db 0100664 root root 0 0 1 X

> dpkg-deb -R clickhouse-client_22.6.2.12_all.deb raw
> cat raw/DEBIAN/md5sums 
2b7327c432fb5acf7066368410358d5d  ./etc/clickhouse-client/config.xml
30469e567fd46902c8124caf71c0e91a  ./usr/share/doc/clickhouse-client/AUTHORS
31b6b824dfd9c33f7ca1a3a81ff733e3  ./usr/share/doc/clickhouse-client/CHANGELOG.md
4ff05e5f8dff9f764e2010427b98b01e  ./usr/share/doc/clickhouse-client/LICENSE
4992af9700ca2d264cc298079f1bb22d  ./usr/share/doc/clickhouse-client/README.md
```

> We can keep extracted binaries inside Docker - the state after package install (to speed up cold start).

Here's another point. Now we have a relatively small image, ~300MB compressed, 948MB raw. And it doesn't contain debug symbols. If we'll use unstriped binaries, as I believe our purpose is, it will be ~1GB compressed and 1.97GB raw.

It's a pretty big difference, AFAIS.

What about trying to discuss it during the sync today?